### PR TITLE
add option to create public neurons when staking

### PR DESCRIPTION
**Pull request by coskucinkilic**
# Motivation

Currently you can create a neuron and change it's visibility to public, however all the new neurons are private by default. We would like to offer simpler way to create public neuron while staking.

# Changes

1. Update nns-service _stakeNeuron_ function to _setNeuronVisibility_ if _asPublicNeuron_ is set to true
2. Update NnsStakeNeuron to display checkbox for public visibility when ENABLE_NEURON_VISIBILITY
3. Add new spacing 'x-small' to Separator to display the form with correct spacing

# Tests
Screenshot 2024-10-21 at 19 07 50

# Todos

- [ ] Add entry to changelog (if necessary).
not necessary

### CODE REVIEW

**Review by mstrasinskis — COMMENTED** [review]
Thank you, looks good overall. Just a couple of minor comments to address.

File: frontend/src/lib/services/neurons.services.ts [hunk 1]
```diff
@@ -228,6 +230,14 @@ export const stakeNeuron = async ({
       fee,
     });
 
+    if (asPublicNeuron) {
```
[the remark this answers is not in the document]
Comment (coskucinkilic) [reply frontend/src/lib/services/neurons.services.ts:233]: I have added the generic failure message for public private neuron change errors which also points them to advance section.

Maybe a specific message for creation would be better.

File: frontend/src/tests/lib/modals/neurons/NnsStakeNeuronModal.spec.ts [hunk 2]
```diff
@@ -328,6 +330,95 @@ describe("NnsStakeNeuronModal", () => {
...
+      it("should not display public neuron checkbox when feature flag is false", async () => {
+        overrideFeatureFlagsStore.setFlag("ENABLE_NEURON_VISIBILITY", false);
+        const po = await renderComponent({});
+
+        expect(
+          await po
+            .getNnsStakeNeuronPo()
+            .getAsPublicNeuronCheckboxPo()
+            .isPresent()
+        ).toBe(false);
```
[the remark this answers is not in the document]
Comment (coskucinkilic) [reply frontend/src/tests/lib/modals/neurons/NnsStakeNeuronModal.spec.ts:348]: Added a single test for creating private neruon when we don't toggle the checkbox without the flag

File: frontend/src/tests/lib/modals/neurons/NnsStakeNeuronModal.spec.ts [hunk 3]
```diff
@@ -328,6 +330,95 @@ describe("NnsStakeNeuronModal", () => {
...
+          await po.getNnsStakeNeuronPo().getAsPublicNeuronCheckboxPo().toggle();
+
+          await po.getNnsStakeNeuronPo().clickCreate();
+
+          expect(stakeNeuron).toBeCalledTimes(1);
+          expect(stakeNeuron).toBeCalledWith({
+            account: mockMainAccount,
+            amount: 22,
+            loadNeuron: true,
+            asPublicNeuron: true,
```
[the remark this answers is not in the document]
Comment (coskucinkilic) [reply frontend/src/tests/lib/modals/neurons/NnsStakeNeuronModal.spec.ts:416]: the new test cover this as well

File: frontend/src/lib/services/neurons.services.ts (hunk 1 above)
[the remark this answers is not in the document]
Comment (mstrasinskis) [reply frontend/src/lib/services/neurons.services.ts:233]: Looks better, thank you.

File: frontend/src/tests/lib/modals/neurons/NnsStakeNeuronModal.spec.ts (hunk 3 above)
[the remark this answers is not in the document]
Comment (mstrasinskis) [reply frontend/src/tests/lib/modals/neurons/NnsStakeNeuronModal.spec.ts:408]: Thanks.

File: frontend/src/tests/lib/services/neurons.services.spec.ts [hunk 4]
```diff
@@ -342,6 +342,62 @@ describe("neurons-services", () => {
...
+        fee: NNS_TOKEN_DATA.fee,
+      });
+      expect(spyStakeNeuron).toBeCalledTimes(1);
+
+      expect(spyChangeNeuronVisibility).toBeCalledWith({
+        neuronIds: [newNeuronId],
+        visibility: NeuronVisibility.Public,
+        identity: mockIdentity,
+      });
+      expect(spyChangeNeuronVisibility).toBeCalledTimes(1);
```
Comment (dskloetd) [inline frontend/src/tests/lib/services/neurons.services.spec.ts:374]: Can you also expect that it's not called when `asPublicNeuron` is false?

File: frontend/src/tests/lib/modals/neurons/NnsStakeNeuronModal.spec.ts [hunk 5]
```diff
@@ -320,6 +322,111 @@ describe("NnsStakeNeuronModal", () => {
       await runResolvedPromises();
       expect(onClose).toBeCalledTimes(1);
     });
+
+    describe("public neuron checkbox", () => {
+      beforeEach(() => {
+        overrideFeatureFlagsStore.reset();
```
[the remark this answers is not in the document]
Comment (coskucinkilic) [reply frontend/src/tests/lib/modals/neurons/NnsStakeNeuronModal.spec.ts:328]: moved the cleanup to top

File: frontend/src/tests/lib/modals/neurons/NnsStakeNeuronModal.spec.ts [hunk 6]
```diff
@@ -320,6 +322,111 @@ describe("NnsStakeNeuronModal", () => {
...
+
+        expect(
+          await po
+            .getNnsStakeNeuronPo()
+            .getAsPublicNeuronCheckboxPo()
+            .isPresent()
+        ).toBe(false);
+      });
+
+      it("should create a private neuron when checkbox is unchecked", async () => {
```
[the remark this answers is not in the document]
Comment (coskucinkilic) [reply frontend/src/tests/lib/modals/neurons/NnsStakeNeuronModal.spec.ts:343]: updated the location of the block, i moved it couple of times and in the end decided that as it acts like feature flag false it should be listed above the block where it is explicitly true, but you are right this is trying to test the condition that is implicitly true.

File: frontend/src/tests/lib/services/neurons.services.spec.ts (hunk 4 above)
Comment (coskucinkilic) [reply frontend/src/tests/lib/services/neurons.services.spec.ts:374]: added

**Review by dskloetd — APPROVED** [review]

**Pull request merged**